File: week02/pset2/readability/readability.c
#include <stdio.h>
#include <string.h>
#include <cs50.h>
#include <ctype.h>
#include <math.h>

// Functions
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float coleman_liau(int letters, int words, int sentences);

int main(void)
{
    // Get the text
    string text = get_string("Text: ");

    // Count the letters
    int num_letters = count_letters(text);

    // Count the words
    int num_words = count_words(text);

    // Count the sentences
    int num_sentences = count_sentences(text);

    float index = coleman_liau(num_letters, num_words, num_sentences);

    // Grade level
    int grade = (int) round(index);
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 15)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }


}

// HELPER FUNCTIONS

// Letter count
int count_letters(string text)
{
    // local variables
    int letter_counter = 0;
    int len_text = strlen(text);

    // count the letters
    for (int i = 0; i < len_text; i++)
    {
        // Only letters!
        if (toupper(text[i]) > 64 && toupper(text[i]) < 91)
        {
            letter_counter++;
        }
    }

    return letter_counter;
}

// Words count
int count_words(string text)
{
    // local variables
    int word_counter = 0;
    int len_text = strlen(text);

    // count the words
    for (int i = 0; i < len_text; i++)
    {
        // Words are separated by spaces!
        if (text[i] == 32)
        {
            word_counter++;
        }
    }

    return word_counter + 1;
}

// Sentences count
int count_sentences(string text)
{
    // local variables
    int sentence_counter = 0;
    int len_text = strlen(text);

    //count the senteces.
    for (int i = 0; i < len_text; i++)
    {
        // Sentences are separated by '.', '!' or, '?'
        if (text[i] == 33 || text[i] == 46 || text[i] == 63)
        {
            sentence_counter++;
        }
    }

    return sentence_counter;
}

// Get the grade of the text
float coleman_liau(int letters, int words, int sentences)
{
    // local variables
    float L, S, index;

    // compute L
    L = (100.0 / words) * letters;

    // compute S
    S = (100.0 / words) * sentences;

    // compute the index
    index = 0.0588 * L - 0.296 * S - 15.8;

    return index;
}

File: week06/pset6/sentimental-readability/readability.py
"""
Usage:
$ python readability.py


Description:
This function grade an input text based on the number of words, letters and sentences
within the text.
The grade is obtained based on the formula of Coleman-Liau:
 - index = 0.0588 * L - 0.296 * S - 15.8


Examples:
$ python readability.py
Text: One fish. Two fish. Red fish. Blue fish.
Before Grade 1

$ python readability.py
Text: Would you like them here or there? I would not like them here or there. I would not like them anywhere.
Grade 2
"""


def main():
    text = input("Text: ")

    # Count letters, words and sentences
    num_letters = count_letters(text)
    num_words = count_words(text)
    num_sentences = count_sentences(text)

    index = int(round(coleman_liau(num_letters, num_words, num_sentences)))

    grade_result = get_grade(index)

    print(grade_result)


# AUXILIAR FUNCTIONS
def count_letters(text):
    """(str)->(int)
    """
    counter = 0
    # only count letters
    for letter in text:
        if letter.isalpha():
            counter += 1

    return counter


def count_words(text):
    """(str)->int
    Every group of letters separated by a space
    count as a word
    """
    # get the length of the list of words in the text
    words = text.split()
    return len(words)


def count_sentences(text):
    """(str)->int
    Every group of words separated by a period,
    exclamation point or question mark indicates the end
    of a sentences.
    """
    sentences = (text.replace("?", ".").replace("!", ".")).split(".")
    return len(sentences) - 1


def coleman_liau(letters, words, sentences):
    """(int, int, int)->float
    """
    # average number of letter per 100 words in the text
    L = (100.0/words) * letters
    # average number of words per 100 words in the text
    S = (100.0/words) * sentences
    # Coleman-Liau index
    return 0.0588 * L - 0.296 * S - 15.8


def get_grade(index):
    """(int)->str
    """
    if index > 16:
        return "Grade 16+"
    elif index < 1:
        return "Before Grade 1"
    else:
        return "Grade " + str(index)


if __name__ == "__main__":
    main()